experimental/gcp-log/function.go

```diff
@@ -23,53 +23,89 @@ import (
+	"github.com/transparency-dev/merkle/rfc6962"
+	"github.com/transparency-dev/serverless-log/pkg/log"
+	"golang.org/x/mod/sumdb/note"
+	"google.golang.org/api/iterator"
 )
 
-func validateCommonArgs(w http.ResponseWriter, origin string) (ok bool, pubKey string) {
-	if len(origin) == 0 {
+type requestData struct {
+	// Common args.
+	Origin         string `json:"origin"`
+	Bucket         string `json:"bucket"`
+	NoteKeyName    string `json:"noteKeyName"`
+	KMSKeyRing     string `json:"kmsKeyRing"`
+	KMSKeyName     string `json:"kmsKeyName"`
+	KMSKeyLocation string `json:"kmsKeyLocation"`
+	KMSKeyVersion  uint   `json:"kmsKeyVersion"`
+
+	// For Sequence requests.
+	EntriesDir string `json:"entriesDir"`
+
+	// For Integrate requests.
+	Initialise bool `json:"initialise"`
+}
+
+func validateCommonArgs(w http.ResponseWriter, d requestData) (ok bool) {
+	if len(d.Origin) == 0 {
 		http.Error(w, "Please set `origin` in HTTP body to log identifier.", http.StatusBadRequest)
-		return false, ""
+		return false
 	}
-
-	pubKey = os.Getenv("SERVERLESS_LOG_PUBLIC_KEY")
-	if len(pubKey) == 0 {
-		http.Error(w,
-			"Please set SERVERLESS_LOG_PUBLIC_KEY environment variable",
+	if len(d.KMSKeyRing) == 0 {
+		http.Error(w, "Please set `kmsKeyRing` in HTTP body to the signing key's key ring.",
+			http.StatusBadRequest)
+		return false
+	}
+	if len(d.KMSKeyName) == 0 {
+		http.Error(w, "Please set `kmsKeyName` in HTTP body to the signing key's name.",
 			http.StatusBadRequest)
-		return false, ""
+		return false
+	}
+	if len(d.KMSKeyLocation) == 0 {
+		http.Error(w, "Please set `kmsKeyLocation` in HTTP body to the signing key's location.",
+			http.StatusBadRequest)
+		return false
+	}
+	if d.KMSKeyVersion == 0 {
+		http.Error(w, "Please set `kmsKeyVersion` in HTTP body to the signing key's version as an integer.",
+			http.StatusBadRequest)
+		return false
+	}
+	if len(d.NoteKeyName) == 0 {
+		http.Error(w, "Please set `noteKeyName` in HTTP body to the key name for the note.",
+			http.StatusBadRequest)
+		return false
 	}
 
-	return true, pubKey
+	return true
 }
 
 // Sequence is the entrypoint of the `sequence` GCF function.
 func Sequence(w http.ResponseWriter, r *http.Request) {
 	// TODO(jayhou): validate that EntriesDir is only touching the log path.
 
-	var d struct {
-		Bucket     string `json:"bucket"`
-		EntriesDir string `json:"entriesDir"`
-		Origin     string `json:"origin"`
-	}
+	// process request args
 
+	d := requestData{}
 	if err := json.NewDecoder(r.Body).Decode(&d); err != nil {
-		code := http.StatusBadRequest
 		fmt.Printf("json.NewDecoder: %v", err)
-		http.Error(w, http.StatusText(code), code)
+		http.Error(w, fmt.Sprintf("Failed to decode JSON: %q", err), http.StatusBadRequest)
 		return
 	}
 
-	ok, pubKey := validateCommonArgs(w, d.Origin)
-	if !ok {
+	if ok := validateCommonArgs(w, d); !ok {
+		return
+	}
+	if len(d.EntriesDir) == 0 {
+		http.Error(w, "Please set `entriesDir` in HTTP body to the key name for the note.",
```

Comment:
I think the error message needs updating here :)

Reply:
Oops, thanks!